Fix NPS Survey Gating

adhiamboperes opened the pull request:
<!-- READ ME FIRST: Please fill in the explanation section below and check off every point from the Essential Checklist! -->
## Explanation
I found out during testing that there is an underlying gating DataProvider update behavior that occurs because of the way I initially designed the survey popup behaviour:
- when the survey popup is displayed, we update the user profile with the current survey shown timestamp, to be used as gating so that the survey isn't triggered again before the grace period ends.
- Updating the profile notifies upstream observers, including the gating DataProvider. 
- Due to this update, the gating provider is triggered to recompute.
- This time, the gating result will be false for triggering the survey, since the grace period condition  has changed.
- The UI behavior is to navigate away from the exploration screen if the survey isn't to be shown. The activity is destroyed, and the popup along with it.

I mitigated this by removing the observer once the initial gating result has been received and processed.

## Tests
I added tests that test survey gating based on changes in the time criterion, while other gating criteria remains constant/fulfilled. Exhaustive gating tests were added in the SurveyGatingController.
 
I also made a change in the `StateFragmentTest` that unregisters the idling resource after initializing profiles instead of at the end of the test runs. Previously, `StateFragmentTest`would not run locally and fail with the error:

```
Error performing 'single click - At Coordinates: 173, 329 and precision: 16, 16' on view 'with id: org.oppia.android:id/play_test_exploration_button'.
at androidx.test.espresso.PerformException$Builder.build(PerformException.java:86)
at androidx.test.espresso.base.DefaultFailureHandler.getUserFriendlyError(DefaultFailureHandler.java:87)
at androidx.test.espresso.base.DefaultFailureHandler.handle(DefaultFailureHandler.java:59)
at androidx.test.espresso.ViewInteraction.waitForAndHandleInteractionResults(ViewInteraction.java:322)
at androidx.test.espresso.ViewInteraction.desugaredPerform(ViewInteraction.java:178)
at androidx.test.espresso.ViewInteraction.perform(ViewInteraction.java:119)
at org.oppia.android.app.player.state.StateFragmentTest.startPlayingExploration(StateFragmentTest.kt:475)org.oppia.android.app.player.state.StateFragmentTest.testFinishChapter_lateNight_isPastGracePeriod_minimumAggregateTimeMet_noSurveyPopup(StateFragmentTest.kt:264)
Caused by: androidx.test.espresso.IdlingResourceTimeoutException: Wait for [TestCoroutineDispatcherIdlingResource] to become idle timed out
at androidx.test.espresso.IdlingPolicy.handleTimeout(IdlingPolicy.java:61)
at androidx.test.espresso.base.UiControllerImpl$5.resourcesHaveTimedOut(UiControllerImpl.java:434)
at androidx.test.espresso.base.IdlingResourceRegistry$Dispatcher.handleTimeout(IdlingResourceRegistry.java:481)
```

This was caused by registering an idling resource before test begins, then when the activity starts, the main thread becomes busy. Generally, the test will continue when the main thread becomes idle, but the idling resource now blocks it and does not call onTransitionToIdle(). So it will always time out. Unregistering the idling resource before the activity starts frees up the main thread.

Disabled running on Robolectric due to [known issue](https://github.com/oppia/oppia-android/issues/1612) with Drag and Drop interaction.

Espresso Run Screenshot
![Screenshot 2024-03-13 at 19 35 12](https://github.com/oppia/oppia-android/assets/59600948/72d8aee9-f39d-4d26-92a9-72fb0919d46c)

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [ ] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
<!-- Delete these section if this PR does not include UI-related changes. -->
If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of the before & after UI changes
- For the screenshots above, include both English and pseudo-localized (RTL) screenshots (see [RTL guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see [accessibility guide](https://github.com/oppia/oppia-android/wiki/Accessibility-A11y-Guide))
- For PRs introducing new UI elements or color changes, both light and dark mode screenshots must be included
- Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing

adhiamboperes committed:
Remove observer after gating result is returned

adhiamboperes committed:
Minor update to syntax

adhiamboperes committed:
fix flaky test

adhiamboperes commented:
@BenHenning, PTAL.

BenHenning commented:
@adhiamboperes is it possible to add tests for any of the discovered user-facing problems?

adhiamboperes committed:
Add survey popup tests

adhiamboperes commented:
> @adhiamboperes is it possible to add tests for any of the discovered user-facing problems?

@BenHenning, I have added some UI tests and updated the PR description.

adhiamboperes committed:
Disable robolectric runs as drag and drop interaction won't work, as documented in https://github.com/oppia/oppia-android/issues/1612

BenHenning commented on app/src/main/java/org/oppia/android/app/player/state/testing/StateFragmentTestActivityPresenter.kt:
```diff
@@ -112,6 +117,46 @@ class StateFragmentTestActivityPresenter @Inject constructor(
     )
   }
 
+  private fun maybeShowSurveyDialog(profileId: ProfileId, topicId: String) {
```
Rather than replicating behavior that exists in ExplorationActivityPresenter could we just verify the specific parts of StateFragment itself that have changed? For instance:
- Ensuring that an exploration completion correctly triggers the survey.
- Ensuring that a data provider change to the survey gate provider doesn't retrigger the survey via StateFragment.

adhiamboperes replied:
Removed this code from the test presenter and added tests for the UI behavior.

BenHenning commented on domain/src/main/java/org/oppia/android/domain/survey/SurveyGatingController.kt:
```diff
@@ -66,7 +66,7 @@ class SurveyGatingController @Inject constructor(
     val showNextTimestamp = lastShownTimestampMs + gracePeriodMillis
 
-    return currentTimeStamp > showNextTimestamp || currentTimeStamp == showNextTimestamp
+    return currentTimeStamp >= showNextTimestamp
```
Any changes to SurveyGatingControllerTest to cover this?

adhiamboperes replied:
While the change here is mainly syntactic and in a private method, I have added some extra test cases to distinguishe between. `==` and `>` conditions.

BenHenning replied:
Thanks. Even behaviors in private methods often can be corresponded to public behaviors (as your tests demonstrate), so it's good to cover them when possible.

BenHenning commented on app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt:
```diff
@@ -528,42 +529,47 @@ class ExplorationActivityPresenter @Inject constructor(
-                .addToBackStack(null)
-                .commit()
-            } else {
+    val liveData = surveyGatingController.maybeShowSurvey(profileId, topicId).toLiveData()
```
Might it be possible to cover these behavioral changes in ExplorationActivityTest and/or ExplorationActivityLocalTest?

adhiamboperes replied:
Added some tests to the ExplorationActivityLocalTest.

BenHenning reviewed (changes requested):
Thanks @adhiamboperes! Left a few comments--PTAL.

adhiamboperes committed:
Additional tests for the SurveyGatingController

adhiamboperes committed:
Remove unused parameter

adhiamboperes committed:
Improve tests

adhiamboperes committed:
Fix failing test

adhiamboperes committed:
Add tests for updating underlying gating dataproviders

adhiamboperes committed:
Add tests for updating underlying gating dataproviders

adhiamboperes committed:
Merge remote-tracking branch 'upstream/fix-survey-gating' into fix-survey-gating

# Conflicts:
#	app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt

adhiamboperes committed:
Fix failing tests

adhiamboperes committed:
Add test for data provider update behaviour

adhiamboperes commented:
@BenHenning, PTAL.

BenHenning commented on app/src/main/java/org/oppia/android/app/player/state/StateFragmentPresenter.kt:
```diff
@@ -569,12 +569,13 @@ class StateFragmentPresenter @Inject constructor(
                   .commitNow()
+
+                // Changes to underlying DataProviders will update the gating result.
+                // liveData.removeObserver(this)
```
If this isn't needed, can the changes be reverted for this presenter?

adhiamboperes replied:
Done

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt:
```diff
@@ -222,20 +223,25 @@ class ExplorationActivityTest {
     testCoroutineDispatchers.registerIdlingResource()
     profileTestHelper.initializeProfiles()
     fakeOppiaClock.setFakeTimeMode(FakeOppiaClock.FakeTimeMode.MODE_UPTIME_MILLIS)
+    testCoroutineDispatchers.unregisterIdlingResource()
```
Why was this changed? Generally the idling resources are needed until the test finishes.

adhiamboperes replied:
Unfortunately, this seems to not work as intended when running the tests on Espresso. This also applies to other large tests like StateFragment Test.

The IdlingResource always times out. From my println logs, I think there is a loop somewhere.

```
java.lang.RuntimeException: java.util.concurrent.ExecutionException: androidx.test.espresso.IdlingResourceTimeoutException: Wait for [TestCoroutineDispatcherIdlingResource] to become idle timed out
at androidx.test.espresso.Espresso.onIdle(Espresso.java:342)
at androidx.test.espresso.Espresso.onIdle(Espresso.java:361)
at org.oppia.android.testing.threading.TestCoroutineDispatchersEspressoImpl.advanceUntilIdle(TestCoroutineDispatchersEspressoImpl.kt:45)
at org.oppia.android.testing.threading.TestCoroutineDispatchersEspressoImpl.runCurrent(TestCoroutineDispatchersEspressoImpl.kt:34)
at org.oppia.android.app.player.exploration.ExplorationActivityTest.testExplorationActivity_loadingAudio_progressbarIsDisplayed(ExplorationActivityTest.kt:1321)
...
androidx.test.rule.ActivityTestRule$ActivityStatement.evaluate(ActivityTestRule.java:527)
at org.oppia.android.testing.OppiaTestRule$apply$1.evaluate(OppiaTestRule.kt:47)
at org.oppia.android.testing.junit.InitializeDefaultLocaleRule$apply$1.evaluate(InitializeDefaultLocaleRule.kt:53)
...
Caused by: java.util.concurrent.ExecutionException: androidx.test.espresso.IdlingResourceTimeoutException: Wait for [TestCoroutineDispatcherIdlingResource] to become idle timed out
at java.util.concurrent.FutureTask.report(FutureTask.java:123)
at java.util.concurrent.FutureTask.get(FutureTask.java:193)
at androidx.test.espresso.Espresso.onIdle(Espresso.java:334)
```

adhiamboperes replied:
I will file a seperate issue to investigate this further

adhiamboperes replied:
Reverting these changes and filing https://github.com/oppia/oppia-android/issues/5379 for investigation instead.

BenHenning reviewed (approved):
Thanks @adhiamboperes! Just a couple of follow-ups. Feel free to merge once they're addressed since the PR otherwise LGTM.

adhiamboperes committed:
Clean up PR

adhiamboperes committed:
Remove redundant var

adhiamboperes committed:
Revert idling resource changes

adhiamboperes committed:
Remove formatting only changes

adhiamboperes commented:
Resolved all comments as fixed and enabling auto merge.

adhiamboperes committed:
Merge branch 'develop' into fix-survey-gating
